### PR TITLE
feat: add CoinMarketCap fetcher

### PR DESCRIPTION
**Pull request by cehali**

### CODE REVIEW

File: packages/oracle-node/src/fetchers/coinmarketcap/CoinMarketCapFetcher.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import axios from "axios";
+import _ from "lodash";
+import { BaseFetcher } from "../BaseFetcher";
+import { getRequiredPropValue } from "../../utils/objects";
+import { FetcherOpts, PricesObj } from "../../types";
+import symbolToId from "./symbol-to-id.json";
+const idToSymbol = _.invert(symbolToId);
+
+const url =
+  "https://pro-api.coinmarketcap.com/v2/cryptocurrency/quotes/latest?id=";
```
Comment (hatskier) [inline packages/oracle-node/src/fetchers/coinmarketcap/CoinMarketCapFetcher.ts:10]: Let's have a pure URL here: https://pro-api.coinmarketcap.com/v2/cryptocurrency/quotes/latest

File: packages/oracle-node/src/fetchers/coinmarketcap/CoinMarketCapFetcher.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    return getRequiredPropValue(symbolToId, symbol);
+  }
+
+  async fetchData(ids: string[], opts: FetcherOpts): Promise<any> {
+    const urlWithIDs = `${url}${ids.join(",")}`;
+    const apiKey = opts.credentials.coinmarketcapApiKey;
+    if (!apiKey) {
+      throw new Error("Missing Coinmarketcap API Key");
+    }
+    const response = await axios.get(urlWithIDs, {
```
Comment (hatskier) [inline packages/oracle-node/src/fetchers/coinmarketcap/CoinMarketCapFetcher.ts:31]: Let's specify ids here using axios request params
```js
const response = await axios.get(urlWithIDs, {
    params: {
       id: ids.join(","),
    },
    headers: {...
});
```

File: packages/oracle-node/src/fetchers/coinmarketcap/CoinMarketCapFetcher.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  override convertIdToSymbol(id: string) {
+    return getRequiredPropValue(idToSymbol, id);
+  }
+
+  override convertSymbolToId(symbol: string) {
+    return getRequiredPropValue(symbolToId, symbol);
+  }
+
+  async fetchData(ids: string[], opts: FetcherOpts): Promise<any> {
+    const urlWithIDs = `${url}${ids.join(",")}`;
```
Comment (hatskier) [inline packages/oracle-node/src/fetchers/coinmarketcap/CoinMarketCapFetcher.ts:26]: It looks too magical 😄

**Review by hatskier — CHANGES_REQUESTED** [review]

File: packages/oracle-node/src/fetchers/coinmarketcap/CoinMarketCapFetcher.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    });
+    return response.data;
+  }
+
+  async extractPrices(response: any, ids: string[]): Promise<PricesObj> {
+    const pricesObj: { [symbol: string]: number } = {};
+    const tokenData = response.data;
+
+    for (const id of ids) {
+      const price = tokenData[id].quote.USD.price;
```
Comment (hatskier) [inline packages/oracle-node/src/fetchers/coinmarketcap/CoinMarketCapFetcher.ts:45]: What do you think about adding try..catch block here? Or just adding `?` operators (smth like `tokenData[id]?.quote?.USD?.price`) to not throw an exception for the whole fetcher when only one symbols is not included in the response

File: packages/oracle-node/src/fetchers/coinmarketcap/example-response.json [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+          "market_cap": 4702317711.3977,
+          "market_cap_dominance": 0.5253,
+          "fully_diluted_market_cap": 11292334011.4,
+          "tvl": null,
+          "last_updated": "2022-11-10T15:23:00.000Z"
+        }
+      }
+    }
+  }
+}
```
Comment (hatskier) [inline packages/oracle-node/src/fetchers/coinmarketcap/example-response.json:549]: How has the packages/oracle-node/src/fetchers/coinmarketcap/symbol-to-id.json file was generated? It is very big and there are no scripts for its generation in this PR

**Review by hatskier — CHANGES_REQUESTED** [review]

File: packages/oracle-node/src/fetchers/coinmarketcap/CoinMarketCapFetcher.ts (hunk 4 above)
Comment (cehali) [reply packages/oracle-node/src/fetchers/coinmarketcap/CoinMarketCapFetcher.ts]: Added handling for this case

File: packages/oracle-node/src/fetchers/coinmarketcap/CoinMarketCapFetcher.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      },
+      headers: {
+        "X-CMC_PRO_API_KEY": apiKey,
+      },
+    });
+    return response.data;
+  }
+
+  async extractPrices(response: any, ids: string[]): Promise<PricesObj> {
+    const pricesObj: { [symbol: string]: number } = {};
```
Comment (hatskier) [inline packages/oracle-node/src/fetchers/coinmarketcap/CoinMarketCapFetcher.ts:41]: Can you replace `symbol` with `id` here?

**Review by hatskier — APPROVED** [review]

**Pull request merged**